lib/spack/spack/directives.py

```diff
@@ -465,23 +465,54 @@ def variant(
         multi (bool): if False only one value per spec is allowed for
             this variant
         validator (callable): optional group validator to enforce additional
-            logic. It receives a tuple of values and should raise an instance
-            of SpackError if the group doesn't meet the additional constraints
+            logic. It receives the package name, the variant name and a tuple
+            of values and should raise an instance of SpackError if the group
+            doesn't meet the additional constraints
+
+    Raises:
+        DirectiveError: if arguments passed to the directive are invalid
     """
     if name in reserved_names:
-        raise ValueError("The variant name '%s' is reserved by Spack" % name)
+        raise DirectiveError(
+            "The variant name '%s' is reserved by Spack" % name
+        )
 
+    # Ensure we have a sequence of allowed variant values, or a
+    # predicate for it.
     if values is None:
-        if default in (True, False) or (type(default) is str and
-                                        default.upper() in ('TRUE', 'FALSE')):
+        if str(default).upper() in ('TRUE', 'FALSE'):
             values = (True, False)
         else:
             values = lambda x: True
 
-    if default is None:
```

Comment:
> I looked through various packages to see if they omitted a default for boolean variants, and I couldn't find an example. It is however a behavior that was supported before by this logic. I didn't see explicit mention of this being supported in the docs, e.g. https://spack.readthedocs.io/en/latest/packaging_guide.html, so it may not be a problem to remove it.

That's what I did too before removing these two lines. In any case, to have `default=False` for a boolean variant you had to specify:
```python
variant('foo', values=(True, False))  # default=False is inferred
```
which is way less clear to me than saying:
```python
variant('foo', default=False)  # The fact that this is a boolean variant is inferred
```

> I'd like to be more-thorough about checking that though (I suppose the unit tests ought to have enforced it if it were a true feature).

This is enforced in the `test_package_sanity` unit test.

Reply:
I'll mark this conversation as resolved, given the "thumbs-up"